source/slang/slang-ast-type.cpp

```diff
@@ -199,41 +95,54 @@ Val* DeclRefType::_substituteImplOverride(ASTBuilder* astBuilder, SubstitutionSe
 {
     if (!subst) return this;
 
-    // the case we especially care about is when this type references a declaration
-    // of a generic parameter, since that is what we might be substituting...
-    if (auto genericTypeParamDecl = as<GenericTypeParamDecl>(declRef.getDecl()))
+    int diff = 0;
+    DeclRef<Decl> substDeclRef = getDeclRef().substituteImpl(astBuilder, subst, &diff);
+
+    // If this declref type is a direct reference to ThisType or a Generic parameter,
+    // and `subst` provides an argument for it, then we should just return that argument.
+    //
+    if (as<DirectDeclRef>(substDeclRef.declRefBase))
```

Comment:
Why is it necessary to check that the decl-ref is a `DirectDeclRef` here?

I know it isn't an issue right now, but if we eventually allowed for generics to take parameters of higher kinds, I could imagine non-direct references to parameters:

```
struct Array<T> { ... }
struct List<T> { ... }

func map< X, Y, C<T> : IContainer>(
    input: C<X>,
    f: X->Y)
    -> C<Y>
{ ... }
```

In that case, we have references to `C<X>` and `C<Y>` that are referring to the parameter as part of a decl-ref that has generic arguments.

I guess if we start to go down that road we'd probably need to fold the substitution that is happening here into the core `substituteImpl` for `DeclRefBase`, at which point we might have to deal with cases where substitution on a decl-ref yields something that isn't a decl-ref (but is still a `Val`)...